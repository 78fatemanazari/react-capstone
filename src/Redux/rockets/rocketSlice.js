import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const url = 'https://api.spacexdata.com/v3/rockets';

const initialState = {
  rockets: [],
  isLoading: true,
  error: undefined,
  hasFetched: false,
};

export const getResultItems = createAsyncThunk('result/getResultItems', async (thunkAPI) => {
  try {
    const resp = await axios(url);
    const myData = resp.data;

    const myTransData = [];

    Object.keys(myData).forEach((key) => {
      const myTmpObj = {
        id: myData[key].id,
        imgPath: myData[key].flickr_images[0],
        rocketName: myData[key].rocket_name,
        active: true,
        description: myData[key].description,
      };
      myTransData.push(myTmpObj);
    });

    return myTransData;
  } catch (error) {
    return thunkAPI.rejectWithValue('something went wrong...');
  }
});

const rocketSlice = createSlice({
  name: 'rocketHandler',
  initialState,
  reducers: {
    reserveRocket: (state, action) => {
      const rocketId = action.payload.id;
      state.rockets = state.rockets.map((rocket) => {
        if (rocket.id === rocketId) {
          return { ...rocket, active: false };
        }
        return rocket;
      });
    },
    cancelRocket: (state, action) => {
      const rocketId = action.payload.id;
      state.rockets = state.rockets.map((rocket) => {
        if (rocket.id === rocketId) {
          return { ...rocket, active: true };
        }
        return rocket;
      });
    },
  },
  extraReducers: {
    [getResultItems.pending]: (state) => {
      state.isLoading = true;
      state.hasFetched = false;
    },
    [getResultItems.rejected]: (state) => {
      state.isLoading = false;
      state.hasFetched = false;
    },
    [getResultItems.fulfilled]: (state, action) => {
      state.isLoading = false;
      state.hasFetched = true;
      state.rockets = action.payload;
    },
  },
});

export const { reserveRocket, cancelRocket } = rocketSlice.actions;
export default rocketSlice.reducer;
