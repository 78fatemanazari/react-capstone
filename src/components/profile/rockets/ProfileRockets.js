import PropTypes from 'prop-types';
import styles from '../../../styles/ProfileRockets.module.css';

const ProfileRockets = ({ rocketProps }) => {
  const myPersonalData = rocketProps.filter((rocket) => rocket.active === false);

  if (myPersonalData.length > 0) {
    return (
      <>
        <ul className="reserved-rocket-list">
          {myPersonalData.map((rocket) => (
            <li key={rocket.id}>{rocket.rocketName}</li>
          ))}
        </ul>
      </>
    );
  }
  return (
    <>
      <div className={styles.noRockets}>
        You do not have any rockets!
      </div>
    </>
  );
};

ProfileRockets.propTypes = {
  rocketProps: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      imgPath: PropTypes.string.isRequired,
      rocketName: PropTypes.string.isRequired,
      active: PropTypes.bool.isRequired,
      description: PropTypes.string.isRequired,
    }),
  ).isRequired,
};

export default ProfileRockets;
