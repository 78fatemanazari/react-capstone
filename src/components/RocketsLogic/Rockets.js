import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import RocketsList from "./RocketsList";
import { getResultItems } from '../redux/rockets/rocketSlice';

const Rockets = () => {
  const { rockets, isLoading, error } = useSelector((store) => store.bookHandler);

  useEffect(() => {
    useDispatch(getResultItems());
  }, []);

  if (isLoading) {
    return <div>Content is loading...</div>
  }

  if (error) {
    return <div>Something went wrong!!!</div>
  }

  return (
    <div>
      <h1>Rockets Page</h1>
      <ul>
        <RocketsList rocketsList={rockets}/>
      </ul>
    </div>
  );
}

export default Rockets;
