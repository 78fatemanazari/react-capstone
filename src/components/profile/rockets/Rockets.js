import { useDispatch, useSelector } from 'react-redux';
import { useEffect } from 'react';
import RocketsList from './RocketsList';
import { getResultItems } from '../../../Redux/rockets/rocketSlice';

import styles from '../../../styles/Rockets.module.css';

const Rockets = () => {
  const dispatch = useDispatch();

  const {
    rockets, isLoading, error, hasFetched,
  } = useSelector((store) => store.rocket);

  useEffect(() => {
    if (!hasFetched) {
      dispatch(getResultItems());
    }
  }, []);

  if (isLoading) {
    return <div>Content is loading...</div>;
  }

  if (error) {
    return <div>Something went wrong!!!</div>;
  }

  return (
    <div>
      <div className="container text-center">
        <hr />
      </div>
      <ul className={styles.listUnstyled}>
        <RocketsList rocketsProps={rockets} />
      </ul>
    </div>
  );
};

export default Rockets;
