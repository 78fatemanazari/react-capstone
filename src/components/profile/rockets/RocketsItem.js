import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserveRocket, cancelRocket } from '../../../Redux/rockets/rocketSlice.js';

import styles from '../../../styles/RocketsItem.module.css';

const RocketsItem = ({
  id, imgPath, rocketName, active, description,
}) => {
  const dispatch = useDispatch();

  return (
    <li className={`container ${styles.element}`}>
      <div className="row">
        <div className="col-md-4 d-flex align-items-center">
          <img alt="rocket_img" src={imgPath} className="img-fluid" />
        </div>
        <div className="col-md-8">
          <h3>
            {rocketName}
          </h3>
          {active ? (
            <p>{description}</p>
          ) : (
            <p className={styles.reserved}>{description}</p>
          )}
          {active ? (
            <button type="button" className={styles.btnReserve} onClick={() => dispatch(reserveRocket({ id }))}>
              Reserve Rocket
            </button>
          ) : (
            <button type="button" className={styles.btnCancel} onClick={() => dispatch(cancelRocket({ id }))}>
              Cancel Reservation
            </button>
          )}
        </div>
      </div>
    </li>
  );
};

RocketsItem.propTypes = {
  id: PropTypes.number.isRequired,
  imgPath: PropTypes.string.isRequired,
  rocketName: PropTypes.string.isRequired,
  active: PropTypes.bool.isRequired,
  description: PropTypes.string.isRequired,
};

export default RocketsItem;
