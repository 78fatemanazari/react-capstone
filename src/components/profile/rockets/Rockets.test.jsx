import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Rockets from './Rockets';
import { getResultItems } from '../../../Redux/rockets/rocketSlice';

const mockStore = configureMockStore([thunk]);

describe('Rockets Componet', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('should display loading state', async () => {
    const store = mockStore({ rocket: { rockets: [], isLoading: true, error: null } });

    const initialState = {
      rockets: [],
      isLoading: true,
      error: undefined,
    };

    render(
      <Provider store={store}>
        <Rockets />
      </Provider>
    );

    store.dispatch(getResultItems());

    expect(screen.getByText('Content is loading...').textContent).toBe('Content is loading...');
  });

  test('should display error state', async () => {
    const store = mockStore({ rocket: { rockets: [], isLoading: false, error: true } });

    const initialState = {
      rockets: [],
      isLoading: true,
      error: undefined,
    };

    render(
      <Provider store={store}>
        <Rockets />
      </Provider>
    );

    store.dispatch(getResultItems());

    expect(screen.getByText('Something went wrong!!!').textContent).toBe('Something went wrong!!!');
  });

  test('should display content', async () => {
    const store = mockStore({ rocket: {
      rockets: [
        {
          id: 1,
          imgPath: 'myImg/path1',
          rocketName: 'first rocket',
          active: true,
          description: 'this is my first rocket description',
        },
        {
          id: 2,
          imgPath: 'myImg/path2',
          rocketName: 'second rocket',
          active: false,
          description: 'this is my second rocket description',
        },
      ],
      isLoading: false,
      error: false,
    } }
  );

    const initialState = {
      rockets: [],
      isLoading: true,
      error: undefined,
    };

    const { rocket } = render(
      <Provider store={store}>
        <Rockets />
      </Provider>
    );

    store.dispatch(getResultItems());

    await waitFor(() => {
      expect(rocket).toMatchSnapshot();
    });
  });
});
