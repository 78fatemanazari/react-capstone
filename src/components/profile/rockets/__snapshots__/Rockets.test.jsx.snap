// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets Componet should display content 1`] = `undefined`;
